File: iosApp/iosApp/AccessTokenViewModel.swift
//
//  AccessTokenViewModel.swift
//  iosApp
//
//  Created by Sergio Fierro on 11/18/20.
//  Copyright © 2020 orgName. All rights reserved.
//

import Foundation
import shared

class AccessTokenViewModel: ObservableObject {
    
    private let accessTokenRepository: AccessTokenRepository
    private let localStorage: LocalStorage
    @Published var url: String {
        didSet {
            localStorage.accessTokenURL = url
        }
    }
    @Published var identity: String = String()
    @Published var accessTokenResponse: AccessTokenResponse? = nil
    
    init(accessTokenRepository: AccessTokenRepository = AccessTokenRepository(), localStorage: LocalStorage = LocalStorage()) {
        self.accessTokenRepository = accessTokenRepository
        self.localStorage = localStorage
        url = self.localStorage.accessTokenURL
    }
    
    func getAccessToken() {
        accessTokenRepository.fetchAccessToken(url: url, identity: identity) { response, error in
            guard let response = response, error == nil else {
                print(error)
                return
            }
            self.accessTokenResponse = response
        }
    }
}

File: androidApp/src/main/java/com/twilio/kmmpoc/androidApp/AccessTokenViewModel.kt
package com.twilio.kmmpoc.androidApp

import androidx.lifecycle.*
import com.twilio.kmmpoc.shared.repository.AccessTokenRepository
import com.twilio.kmmpoc.shared.repository.AccessTokenResponse
import com.twilio.kmmpoc.shared.repository.LocalStorage
import kotlinx.coroutines.Dispatchers

class AccessTokenViewModel(private val accessTokenRepository: AccessTokenRepository, private val localStorage: LocalStorage): ViewModel() {

    private val identity: MutableLiveData<String> = MutableLiveData()
    val accessToken: LiveData<AccessTokenResponse> = identity.switchMap {
        liveData(context = viewModelScope.coroutineContext + Dispatchers.IO) {
            emit(accessTokenRepository.fetchAccessToken(url.value ?: "", it))
        }
    }

    private val _url: MutableLiveData<String> = MutableLiveData()
    val url: LiveData<String> = _url

    init {
        setURL(localStorage.accessTokenURL)
    }

    fun setURL(url: String) {
        localStorage.accessTokenURL = url
        _url.value = url
    }

    fun getAccessToken(identity: String) {
        this.identity.value = identity
    }
}